### PR TITLE
fix: cache per command rather than per 'fs' instance

### PR DESCRIPTION
**Pull request by wmhilton**
TLDR: If you're doing a bunch of git commands in a loop, you might see a performance regression. File an issue letting me know what you're doing in a loop, and I'll figure out how to speed it up again.

Since some commit a while ago, we've been caching the parsed `.git/index` data since it's needed constantly during `walk`. But we were caching it keyed by `fs` and `filepath`. And... well, in situations where `fs` is in the browser or in memory, and you do funny things with it like wipe it or switch out the files for a different set of files, then `fs` isn't very reliable as a caching key. Because even though the `fs` object reference might be the same, it could be holding a very different file system.

To fix this, I'm relaxing the caching so we only hold onto the cached `.git/index` for the duration of the current command. If you're doing a bunch of `git` commands in a loop, you might see a performance regression. File an issue letting me know what you're doing in a loop, and I'll figure out how to speed it up again.

### CODE REVIEW

**Comment by karma-pr-reporter** [comment]
## Test Results for 584487894fde905cf68d01287614575f6f34a2ed:
| Browser                                 | Pass | Skip | Fail | Time   | Disconnect? |
| --------------------------------------- | ---- | ---- | ---- | ------ | ----------- |
| Chrome 79.0.3945 (Windows 10 0.0.0)     | 362  | 0    | 1    | 2m:42s | false       |
| Chrome Mobile 80.0.3987 (Android 0.0.0) | 363  | 0    | 0    | 2m:7s  | false       |
| Firefox 74.0.0 (Ubuntu 0.0.0)           | 363  | 0    | 0    | 1m:21s | false       |
| HeadlessChrome 0.0.0 (Linux 0.0.0)      | 363  | 0    | 0    | 1m:3s  | false       |
| Mobile Safari 13.0.0 (iOS 13.0.0)       | 363  | 0    | 0    | 3m:4s  | false       |
| Safari 13.0.3 (Mac OS X 10.15.1)        | 363  | 0    | 0    | 3m:39s | false       |

### Erroring Tests

- Chrome 79.0.3945 (Windows 10 0.0.0):
  - annotatedTag creates a signed tag to HEAD

### Longest running tests
- 15.875 secs - clone clone with noTags - _Firefox_
- 12.113 secs - clone clone with noTags - _Mobile Safari_
- 11.234 secs - clone clone with noTags - _Safari_
- 11.039 secs - clone clone with noTags - _HeadlessChrome_
- 10.858 secs - clone clone with noTags - _Chrome_
- 8.34 secs - walk can populate type, mode, oid, and content - _Safari_
- 7.965 secs - fetch fetch --prune from git-http-mock-server - _Safari_
- 6.897 secs - fetch fetch --prune-tags from git-http-mock-server - _Chrome_
- 6.39 secs - statusMatrix statusMatrix with filter - _Safari_
- 6.362 secs - fetch fetch (from Github) - _Firefox_
- 6.309 secs - clone clone with noTags - _Chrome Mobile_

**Comment by karma-pr-reporter** [comment]
## Test Results for 584487894fde905cf68d01287614575f6f34a2ed:
| Browser                                 | Pass | Skip | Fail | Time   | Disconnect? |
| --------------------------------------- | ---- | ---- | ---- | ------ | ----------- |
| Chrome 79.0.3945 (Windows 10 0.0.0)     | 363  | 0    | 0    | 2m:21s | false       |
| Chrome Mobile 80.0.3987 (Android 0.0.0) | 363  | 0    | 0    | 2m:11s | false       |

### Longest running tests
- 10.321 secs - clone clone with noTags - _Chrome_
- 9.531 secs - clone clone with noTags - _Chrome Mobile_
- 4.122 secs - Hosting Providers AWS CodeCommit fetch - _Chrome Mobile_
- 4.104 secs - Hosting Providers AWS CodeCommit push - _Chrome Mobile_
- 4.075 secs - fetch fetch --prune from git-http-mock-server - _Chrome Mobile_
- 4.07 secs - fetch fetch --prune from git-http-mock-server - _Chrome_
- 4.066 secs - fetch shallow fetch (from Github) - _Chrome Mobile_
- 3.854 secs - fetch shallow fetch (from Github) - _Chrome_
- 3.759 secs - walk can populate type, mode, oid, and content - _Chrome Mobile_
- 3.682 secs - statusMatrix statusMatrix with filepaths - _Chrome_
- 3.416 secs - Hosting Providers GitHub push - _Chrome_
